pkg/resources/grant_database_role.go

```diff
@@ -0,0 +1,228 @@
+		if err := client.DatabaseRoles.Grant(ctx, req); err != nil {
+			return err
+		}
+	} else if parentDatabaseRoleName, ok := d.GetOk("parent_database_role_name"); ok && parentDatabaseRoleName.(string) != "" {
+		parentRoleIdentifier := sdk.NewDatabaseObjectIdentifierFromFullyQualifiedName(parentDatabaseRoleName.(string))
+		snowflakeResourceID = helpers.EncodeSnowflakeID(databaseRoleIdentifier.FullyQualifiedName(), sdk.ObjectTypeDatabaseRole.String(), parentRoleIdentifier.FullyQualifiedName())
+		req := sdk.NewGrantDatabaseRoleRequest(databaseRoleIdentifier).WithDatabaseRole(parentRoleIdentifier)
+		if err := client.DatabaseRoles.Grant(ctx, req); err != nil {
+			return err
+		}
+	} else if shareName, ok := d.GetOk("share_name"); ok && shareName.(string) != "" {
+		shareIdentifier := sdk.NewAccountObjectIdentifierFromFullyQualifiedName(shareName.(string))
+		snowflakeResourceID = helpers.EncodeSnowflakeID(databaseRoleIdentifier.FullyQualifiedName(), sdk.ObjectTypeShare.String(), shareIdentifier.FullyQualifiedName())
+		req := sdk.NewGrantDatabaseRoleToShareRequest(databaseRoleIdentifier, shareIdentifier)
+		if err := client.DatabaseRoles.GrantToShare(ctx, req); err != nil {
+			return err
+		}
+	}
+	d.SetId(snowflakeResourceID)
+	return ReadGrantDatabaseRole(d, meta)
+}
+
+// ReadGrantDatabaseRole implements schema.ReadFunc.
+func ReadGrantDatabaseRole(d *schema.ResourceData, meta interface{}) error {
+	db := meta.(*sql.DB)
+	client := sdk.NewClientFromDB(db)
+	parts := strings.Split(d.Id(), helpers.IDDelimiter)
+	databaseRoleName := parts[0]
+	databaseRoleIdentifier := sdk.NewDatabaseObjectIdentifierFromFullyQualifiedName(databaseRoleName)
+	objectType := parts[1]
+	targetIdentifier := parts[2]
+	ctx := context.Background()
+	grants, err := client.Grants.Show(ctx, &sdk.ShowGrantOptions{
+		Of: &sdk.ShowGrantsOf{
+			DatabaseRole: databaseRoleIdentifier,
+		},
+	})
+	if err != nil {
+		log.Printf("[DEBUG] database role (%s) not found", databaseRoleIdentifier.FullyQualifiedName())
+		d.SetId("")
+		return nil
+	}
+
+	var found bool
+	for _, grant := range grants {
+		if grant.GrantedTo == sdk.ObjectType(objectType) {
+			if grant.GrantedTo == sdk.ObjectTypeRole || grant.GrantedTo == sdk.ObjectTypeShare {
+				if grant.GranteeName.FullyQualifiedName() == targetIdentifier {
+					found = true
+					break
+				}
+			} else {
+				/*
+					note that grantee_name is not saved as a valid identifier in the
+					SHOW GRANTS OF DATABASE ROLE <database_role_name> command
+					for example, "ABC"."test_parent_role" is saved as ABC."test_parent_role"
+					or "ABC"."test_parent_role" is saved as ABC.test_parent_role
+					and our internal mapper thereby fails to parse it correctly, returning "ABC."test_parent_role"
+					so this funny string replacement is needed to make it work
+				*/
+				s := grant.GranteeName.FullyQualifiedName()
+				if !strings.Contains(s, "\"") {
```

Comment:
But this will fail if the dot is included in the quoted part, right? (like `"one"."tw.o"`).

This is a broader topic and I will include it in the upcoming identifiers rework.

I also believe that it should be handled on the SDK level where the grant row is converted to our Grant. That way we won't have to deal with that in every resource using it. I don't expect it to be moved necessarily in this PR (it would be nice to have). Let's at least extract it to some helper and unit test it thoroughly.

Reply:
Yeah, you're right, random dots in the identifier name can make it divide in the wrong places :/ but yeah there should be one way of handling identifier parsing that maybe will come soon. Resolving the comment.